src/Symfony/Component/DomCrawler/Crawler.php

```diff
@@ -593,6 +593,33 @@ public function html()
     }
 
     /**
+     * Evaluates an XPath expression.
+     *
+     * Since an XPath expression might evaluate to either a simple type or a \DOMDoneList,
+     * this method will return either an array of simple types or a new Crawler instance.
+     *
+     * @param string $xpath An XPath expression
+     *
+     * @return array|Crawler An array of evaluation results or a new Crawler instance
+     */
+    public function evaluate($xpath)
+    {
+        $data = array();
+        $prefixes = $this->findNamespacePrefixes($xpath);
+
+        foreach ($this->nodes as $node) {
+            $domxpath = $this->createDOMXPath($node->ownerDocument, $prefixes);
```

Comment:
you can use a single DOMXPath now. As of Symfony 3, it is impossible to have nodes belonging to different documents. They will all belong to ``$this->document`` (which can only be ``null`` for an empty root crawler)